### DISCUSSION TITLE
sympy.physics.mechanics

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#24647] [Q&A]
As I do not know, where to put this, I put it here.

I am trying to model something simple: a disc running on an uneven street, all in 2D.
I modelled it the way I learned to use sympy.physics.mechanics. The motions look reasonable, but I cannot get the total energy to be constant, except when the street is completely flat.
I have been on and off this system for a few months now, but I do not manage to achieve any progress.

I would of course like to know, where my modelling error is.
For anyone interested, I have added my code below.
I believe, my comments explain, what I am doing, but of course, I have looked at it so often myself....
Any help is greatly appreciated!

```

.. code:: ipython3

    import sympy as sm
    import sympy.physics.mechanics as me
    import numpy as np
    from scipy.integrate import solve_ivp
    from scipy.optimize import fsolve, minimize
    
    from matplotlib import animation
    import matplotlib
    from matplotlib import patches
    import matplotlib.pyplot as plt
    from IPython.display import HTML
    matplotlib.rcParams['animation.embed_limit'] = 2**128
    %matplotlib inline
    
    import time

A homogeneous disc with radius r and mass m is running on an uneven 'street' without sliding. 
The disc is not allowed to jump, hence I can calculate the reaction forces needed to hold it on the
street at all times

The 'overall' shape of the street is modelled as a parabola (called strassen_form), the unevenness is 
modelled as a sum of sin functions (called strasse) with each term having a smaller amplitude and higher 
frequency as the previous one.
the 'street itself' is the sum of strasse and strassen_form, called gesamt.

NOTE: The disc must always have only have one contact point, hence this inequality 
must hold: $r < | \dfrac{(1 + (d/dx(gesamt(x(t))))^{3/2}} {d^2/dx^2(gesamt(x(t)))}|$, 
formula for osculating radius (Krümmungsradius) from Wikipedia.

Using *simplify()* on some intermediate results reduces the number of operations substantially.


**Variables**

- $q_1, u_1$: rotation of the disc and its speed
- $x, u_x$: loction of the contact point and the spped of successive contact points

- N: inertial frame
- $P_0$ point fixed in N
- $A_2$ body fixed frame of disc
- $Dmc$: center of disc
- $Dmc_o$: location of observer

- $m$: mass of the disc
- $m_o$: mass of the observer
- $i_{ZZ}$: moment of inertia of the disc aroud the Z axis
- $\alpha, \beta$: location of the observer relative to the center of the disc
- $amplitude, frequenz$: parameters of the stree
- $reibung$: friction


.. code:: ipython3

    start = time.time()
    
    # q1 is the angle of the disc, x is the horizontal position of contact point CP
    q1, x = me.dynamicsymbols('q1 x')  
    u1, ux = me.dynamicsymbols('u1 ux')
    
    auxx, auxy, fx, fy = me.dynamicsymbols('auxx auxy fx fy')   # for the reaction forces at contact point CP
    
    m, mo, g, r, iZZ, alpha, beta = sm.symbols('m, mo, g, r, iZZ, alpha, beta')
    amplitude, frequenz, reibung, t = sm.symbols('amplitude frequenz reibung t') # reibung = friction in German
    
    N = me.ReferenceFrame('N')         # fixed inertial frame
    A2 = me.ReferenceFrame('A2')       # fixed to the disc
    P0, CP, Dmc, Dmco = sm.symbols('P0, CP, Dmc, Dmco', cls=me.Point)

Determine the street and its osculating radius (Schmiegekreis). As the disc should contact the street at exactly one point, this osculating radius must be larger than the radius of the disc. I found the formula for the osculating radius of a function in the internet.


.. code:: ipython3

    #============================================
    rumpel = 5    # the higher the number the more 'uneven the street'
    #============================================
    strasse = sum([amplitude/j * sm.sin(j*frequenz * x) for j in range(1, rumpel)])
    strassen_form = (frequenz/1. * x)**2
    gesamt = strassen_form + strasse
    
    r_max = (sm.S(1.) + (gesamt.diff(x))**2 )**sm.S(3/2)/gesamt.diff(x, 2)

**Relationship of x(t) to q(t)**:

Obviously, $ x(t) = function(q(t), gesamt(x(t), r) $.
When the disc is rotated through an angle $q$, the arc length is $r * q(t)$.

The arc length of a function f(k(t)) from 0 to $x(t)$ is: $ \int_{0}^{x(t)} \sqrt{1 +  (d/dx(f(k(t))^2} \,dk \ $ (I found this in the internet)

This gives the sought after relationship between $q(t)$ and $x(t)$:
$ r \cdot (-q(t))  =  \int_{0}^{x(t)} \sqrt{1 + (d/dk(gesamt(k(t))^2}\,dk \ $, differentiated w.r.t *t*:
- $ r \cdot (-u)  = \sqrt{1 + (d/dx(gesamt(x(t))^2} \cdot d/dt(x(t)) $, that is solved for $d/dt(x(t))$:

- $ d/dt(x(t)) = \dfrac{-(r \cdot u)} {\sqrt{1 + (d/dx(gesamt(x(t))^2}}$

The - sign is a consequence of the 'right hand rule' for frames. This is the sought after first order differential equation for $x(t)$.

.. code:: ipython3

    rhs3 = (-u1 * r / sm.sqrt(1. + (gesamt.diff(x))**2)).simplify()

The vector perpendicular to the strasse is -(gesamt.diff(x), - 1). I found this in the internet.\
The leading minus sign, because directed 'inward'. It points from the contact point CP to the geometric center of the disc Dmc. 

As the contact point is (momentarily) fixed in A2, I can use **v2pt_theory** to get the speed of Dmc

.. code:: ipython3

    vector = (-(gesamt.diff(x)*N.x - N.y)).simplify()
    
    A2.orient_axis(N, q1, N.z)
    A2.set_ang_vel(N, u1 * N.z)
    
    CP.set_pos(P0, x*N.x + gesamt*N.y)    # location of contact point
    # CP has no real velocity, as it is 'part of the street' auxx, auxy are virtual speeds to determine the reaction forces on CP
    CP.set_vel(N, auxx*N.x + auxy*N.y)# + CP.pos_from(P0).diff(t, N).subs({sm.Derivative(x, t): rhs3}))
    #The center of the wheel is at distance r from CP, perpendicular to the surface of the street.
    Dmc.set_pos(CP, r * (vector.normalize()).simplify())
    Dmco.set_pos(Dmc, r * (alpha*A2.x + beta*A2.y))
    Dmc.v2pt_theory(CP, N, A2)
    Dmco.v2pt_theory(Dmc, N, A2)
    
    # just needed for potting later
    Dmco_pos = [me.dot(Dmco.pos_from(P0), uv) for uv in (N.x, N.y)]
    Dmc_pos = [me.dot(Dmc.pos_from(P0), uv) for uv in (N.x, N.y)]
    CP_pos = [me.dot(CP.pos_from(P0), uv) for uv in (N.x, N.y)] # just for the plotting

Define the bodies and the energies.
I find it useful to calculate and plot the energies: They may give hints, that there may be a mistake in setting up Kane's equations.

.. code:: ipython3

    I = me.inertia(A2, 0., 0., iZZ)                                              
    Body = me.RigidBody('Body', Dmc, A2, m, (I, Dmc))
    observer = me.Particle('observer', Dmco, mo)
    BODY = [Body, observer]
    kin_energie = (Body.kinetic_energy(N) + observer.kinetic_energy(N)).subs({auxx: 0., auxy:  0.})
    pot_energie = m * g * me.dot(Dmc.pos_from(P0), N.y) + mo * g * me.dot(Dmco.pos_from(P0), N.y) 
    
    print('kinetic energy DS: ', me.find_dynamicsymbols(kin_energie))
    print('kinetic energy free symbols', kin_energie.free_symbols)
    print('potential energy DS: ', me.find_dynamicsymbols(pot_energie))
    print('potential energy free symbols', pot_energie.free_symbols)

**Kane's equations**

- determine the external forces, here only gravitational forces
- get the equation for the reaction forces. They (of course) depend on the accelerations of the masses, hence on $rhs = MM^{-1} * force$. As this is a small system, I calculate *rhs* symbolically and use it. for larger systems this is too slow, better done numerically.
- add the term to calculate the  X - position of CP at the bottom of force. Recall we got a differential equation for x(t).
- enlarge the mass matrix appropriately
- convert the sympy functions to numpy functions, using *lambdify(..)*

.. code:: ipython3

    FL = [(Dmc, -m*g*N.y), (Dmco, -mo*g*N.y), (CP, fx*N.x + fy*N.y), (A2, -reibung*u1*A2.z)]
    kd = [u1 - q1.diff(t)]
    q = [q1]
    u = [u1]
    aux = [auxx, auxy]
    
    KM = me.KanesMethod(N, q_ind=q, u_ind=u, kd_eqs=kd, u_auxiliary=aux)
    (fr, frstar) = KM.kanes_equations(BODY, FL)
    MM = KM.mass_matrix_full
    force = KM.forcing_full
    
    rhs = KM.rhs().subs({sm.Derivative(x, t): rhs3})
    print('rhs DS', me.find_dynamicsymbols(rhs))
    print('rhs free symbols', rhs.free_symbols)
    print('rhs has {} operations'.format(sum([rhs[i].count_ops(visual=False) for i in range(len(rhs))])), '\n')
    
    # Reaction forces
    eingepraegt = KM.auxiliary_eqs.subs({sm.Derivative(u1, t): rhs[1], sm.Derivative(x, t): rhs3})
    print('eingepraegt DS', me.find_dynamicsymbols(eingepraegt))
    print('eingepraegt free symbols', eingepraegt.free_symbols)
    print('eingepraegt has {} operations'.format(sum([eingepraegt[i].count_ops(visual=False) for i in range(len(eingepraegt))])), '\n')
    
    # Add rhs3 at the bottom of force, to get d/dt(x) = rhs3. This is to numerically integrate x(t)
    force = sm.Matrix.vstack(force, sm.Matrix([rhs3])).subs({sm.Derivative(x, t): rhs3, fx: 0., fy: 0.}) 
    print('force DS', me.find_dynamicsymbols(force))
    print('force free symbols', force.free_symbols)
    print('force has {} operations'.format(sum([force[i].count_ops(visual=False) for i in range(len(force))])), '\n')
    
    # Enlarge MM properly
    MM = sm.Matrix.hstack(MM, sm.Matrix([0., 0.])).subs({sm.Derivative(x, t): rhs3})
    MM = sm.Matrix.vstack(MM, sm.Matrix([0., 0., 1.]).T)
    print('MM DS', me.find_dynamicsymbols(MM))
    print('MM free symbols', MM.free_symbols)
    print('MM has {} operations'.format(sum([MM[i, j].count_ops(visual=False) for i in range(MM.shape[0]) for j in range(MM.shape[1])])), '\n')
    
    # Lambdification. Turning symbolic expressions into numpy functions.
    pL = [m, mo, g, r, iZZ, alpha, beta, amplitude, frequenz, reibung]
    qL = q + u + [x]
    F = [fx, fy]
    
    MM_lam = sm.lambdify(qL + pL, MM, cse=True)
    force_lam = sm.lambdify(qL + pL, force, cse=True)
    
    CP_pos_lam = sm.lambdify(qL + pL, CP_pos, cse=True)
    Dmc_pos_lam = sm.lambdify(qL + pL, Dmc_pos, cse=True)
    Dmco_pos_lam = sm.lambdify(qL + pL, Dmco_pos, cse=True)
    
    # will be solved for F numerically later. Much too large to be solved symbollically.
    eingepraegt_lam = sm.lambdify(F + qL + pL, eingepraegt, cse=False) 
    
    #this is needed to plot the shape of the street
    strasse_lam = sm.lambdify([x] + pL,  gesamt, cse=True)
    
    kin_lam = sm.lambdify(qL + pL, kin_energie, cse=True)
    pot_lam = sm.lambdify(qL + pL, pot_energie, cse=True)
    
    r_max_lam = sm.lambdify([x] + pL, r_max, cse=True)
    
    print('it took {:.3f} sec to establish Kanes equations'.format(time.time() - start))

**Numerical integration**

Parameter / initial values
- $q_{11}, u_{11}$: rotation of the disc and its speed
- $x_1, u_{x1}$: loction of the contact point and the spped of successive contact points

- $m_1$: mass of the disc
- $m_{o1}$: mass of the observer
- $i_{ZZ1}$: moment of inertia of the disc aroud the Z axis
- $\alpha_1, \beta_1$: location of the observer relative to the center of the disc
- $amplitude_1, frequenz_1$: parameters of the street
- $reibung_1$: friction

While it makes sense to call these values similar to their names when setting up Kane's equations, **avoid** the **same** name: The symbols / dynamic symbols get overwritten, with unintended consequences.

.. code:: ipython3

    start1 = time.time()
    
    # Input parameters 
    #==========================================================
    m1 = 1.
    mo1 = 1.
    r1 = 1.475
    alpha1, beta1 = 0., 0.95
    amplitude1 = 1.
    frequenz1 = 0.25     # the smaller this number, the more 'even' the street   
    reibung1 = 0.        # Friction
    intervall = 15.     # time inverval of integration is [0., intervall]
    schritte = 500      # Where the results of solve_ivp will be given, see description of solve_ivp.
    
    q11 = 0.             # starting angle. As the disc is symmetric about this angle, it plays no real role
    u11 = 5.             # starting angular velocity of disc.
    x1 = 10.             # Starting X position of disc. 
    #==========================================================
    iZZ1 = 1/2 * m1 * r1**2
    pL_vals = [m1, mo1,  9.8, r1, iZZ1, alpha1, beta1, amplitude1, frequenz1, reibung1]
    y0 = [q11, u11, x1]
    print('Arguments')
    print('[m, mo1,  g, r, iZZ, iXY, alpha, beta, amplitude, frequenz, reibung]')
    print(pL_vals, '\n')
    print('[q11, u11, x1]')
    print(y0, '\n')
    
    startwert = y0[2]   # just needed for the plots below
    startomega = y0[1]  #  dto.
    
    #find the largest admissible r, given strasse, amplitude, frequenz
    def func(x, args):
    # just needed to get the arguments matching for minimize
        return np.abs(r_max_lam(x, *args))
    
    x0 = 0.1            # initial guess
    minimal = minimize(func, x0, pL_vals)
    
    if pL_vals[3] < (x111 := minimal.get('fun')):
        print('selected radius = {} is less than maximally admissible radius = {:.2f}, hence o.k.'.format(pL_vals[3], x111), '\n')
    else:
        print('selected radius {} is larger than admissible radius {:.2f}, hence NOT o.k.'.format(pL_vals[3], x111), '\n')
     
    times = np.linspace(0, intervall, schritte)
    
    def gradient(t, y, args):
        vals = np.concatenate((y, args))
        sol = np.linalg.solve(MM_lam(*vals), force_lam(*vals))
        return np.array(sol).T[0]
    
    # method = 'Radau' seems to work best here.
    resultat1 = solve_ivp(gradient, (0., float(intervall)), y0, t_eval=times, args=(pL_vals,), 
        atol=1.e-6, rtol=1.e-6)#, method='Radau')
    resultat = resultat1.y.T
    event_dict = {-1: 'Integration failed. Do not run the plot', 0: 'Integration finished successfully', 1: 'some termination event'}
    print(event_dict[resultat1.status], ' message is:', resultat1.message)
    print('resultat shape', resultat.shape, '\n')
    print("To numerically integrate an intervall of {} sec the routine cycled {} times and it took {:.5f} sec ".format(intervall, resultat1.nfev, time.time() - start1))

Plot the generalized coordinates and speeds

.. code:: ipython3

    Dmc_X = np.empty(schritte)
    Dmc_Y =np.empty(schritte)
    for i in range(schritte):
        Dmc_X[i], Dmc_Y[i] = Dmc_pos_lam(*[resultat[i, j] for j in range(resultat.shape[1])], *pL_vals)
    
    fig, ax = plt.subplots(figsize=(10, 5))
    for i, j in zip(range((resultat.shape[1])), ('rotational angle', 'rotational speed', 'displacement')):
        ax.plot(times, resultat[:, i], label=j)
    ax.set_title('Coordinates, friction = {}'.format(reibung1))
    ax.legend();

Plot the reaction forces 

.. code:: ipython3

    def func (x11, *args):
    # just serves to make the arguments compatible between fsolve and eingepraegt_lam
        return eingepraegt_lam(*x11, *args).reshape(len(F))
    
    kraft = np.empty((schritte, len(F)))
    x0 = tuple([1. for i in range(len(F))])   # initial guess
    for i in range(schritte):
        y00 = [resultat[i, j] for j in range(resultat.shape[1])]
        args = tuple((y00 + pL_vals))
        A = fsolve(func, x0, args=args).reshape(len(F)) # numerically find fx, fy
        x0 = tuple(A)      # updated initial guess, should improve convergence
        kraft[i] = A        
            
    fig, ax = plt.subplots(figsize=(10, 5))
    ax.plot(times, kraft[:, 0], label = 'Fx')
    ax.plot(times, kraft[:, 1], label = 'Fy')
    ax.set_title('Reaction forces on contact point, friction = {}'.format(reibung1))
    ax.legend();

Plot the energies of the system. Absent any friction the total energy should be constant. It is not, unless the osculating circle is much larger than the radius of the disc.\
I do not know, where the error comes from. It does not seem to be numerical, changing $a_{tol}, r_{tol}$ or the *method* does not seem to have any effect. But I cannot find an error in the description of the systen, either.\
If the street is totally flat (*amplitude* = *frequenz* = 0.) the total energy is constant. 

.. code:: ipython3

    kin_np = np.empty(schritte)
    pot_np = np.empty(schritte)
    total_np = np.empty(schritte)
    
    for i in range(schritte):
        kin_np[i] = kin_lam(*[resultat[i, j] for j in range(resultat.shape[1])], *pL_vals)
        pot_np[i] = pot_lam(*[resultat[i, j] for j in range(resultat.shape[1])], *pL_vals)
        total_np[i] = kin_np[i] + pot_np[i]
    if pL_vals[-1] == 0.:
        print('Max deviation from constant of total energy is {:.4f} % of max total energy'
              .format((max(total_np) - min(total_np))/max(total_np) * 100.))
        print('Max absolute deviation from constant of total energy is {:.4f} Nm'
              .format((np.max(total_np) - np.min(total_np))))
    fig, ax = plt. subplots(figsize=(10, 5))
    ax.plot(times, kin_np, label='kinetic energy')
    ax.plot(times, pot_np, label='pos energy')
    ax.plot(times, total_np, label='total energy')
    ax.set_title('Energy of the disc, friction = {}'.format(reibung1))
    ax.legend();

Plot the street and the extreme positions of the disc

.. code:: ipython3

    # plot the street, and the extremes, of the position of the disc.
    fig, ax = plt.subplots(figsize=(10, 5))
    links = np.min(resultat[:, 2])
    rechts = np.max(resultat[:, 2])
    ruhe = np.mean([resultat[-30::, 2]])    # get approx. rest position of wheel
    maximal = max(np.abs(links), np.max(rechts))
    times1 = np.linspace(-maximal-5, maximal+5, schritte)
    ax.plot(times1, strasse_lam(times1, *pL_vals)  , label='Strasse')
    if pL_vals[-1] != 0.:
        ax.axvline(ruhe,ls = '--', color='red', label='approx. fimal pos. of wheel')
    ax.axvline(links,ls = '--', color='green', label='leftmost pos. of wheel')
    ax.axvline(rechts,ls = '--', color='black', label='rightmost pos. of wheel');
    ax.axvline(startwert, ls='--', color='orange', label='starting position of wheel')
    if startomega > 0.:
        richtung = 'left'
    else:
        richtung = 'right'
    text = 'Wheel has speed ' + str(np.abs(startomega)) + ' units to the ' + richtung
    plt.title(text + ', friction = {} '.format(reibung1))
    ax.legend();

**Animation**

The blue dot represents the observer.
HTML(...) is needed for the animation to show on an iPad, no idea, whether needed on other machines. It is SLOW!

.. code:: ipython3

    #Animation
    # Location of the center of the disc
    Dmcx = np.empty(schritte)
    Dmcy =np.empty(schritte)
    Dmcox = np.empty(schritte)
    Dmcoy =np.empty(schritte)
    
    for i in range(schritte):
        Dmcx[i], Dmcy[i] = Dmc_pos_lam(*[resultat[i, j] for j in range(resultat.shape[1])], *pL_vals)
        Dmcox[i], Dmcoy[i] = Dmco_pos_lam(*[resultat[i, j] for j in range(resultat.shape[1])], *pL_vals)
    
    # needed to give the picture the right size.
    xmin = min([resultat[i, 2] for i in range(schritte)])
    xmax = max([resultat[i, 2] for i in range(schritte)])
    
    ymin = min([strasse_lam(resultat[i, 2], *pL_vals) for i in range(schritte)]) 
    ymax = max([strasse_lam(resultat[i, 2], *pL_vals) for i in range(schritte)]) 
    
    # Data to draw the uneven street
    cc = r1
    strassex = np.linspace(xmin - 3*cc, xmax + 3.*cc, schritte)
    strassey = [strasse_lam(strassex[i], *pL_vals) for i in range(len(strassex))]
    
    # Data for the dashed lines, which mark the contact point of the ellipse with the street
    geradex = np.linspace(xmin - 3.*cc, xmax + 3.*cc, schritte)
    geradey = np.linspace(ymin - 3.*cc, ymax + 3.*cc, schritte)
    
    vertikal = np.empty(schritte)
    horizontal =np.empty(schritte)
    vertikal_store = []
    horizontal_store = []
    for i in range(schritte):
        vertikal = np.array([resultat[i, 2] for k in range(schritte)])
        vertikal_store.append(vertikal)
        horizontal = np.array([strasse_lam(resultat[i, 2], *pL_vals) for k in range(schritte)])
        horizontal_store.append(horizontal)
    
    def animate_pendulum(times, x1, y1, x2, y2):    
        fig, ax = plt.subplots(figsize=(10, 10), subplot_kw={'aspect': 'equal'})
        
        ax.axis('on')
        ax.set_xlim(xmin - 3.*cc, xmax + 3.*cc)
        ax.set_ylim(ymin - 3.*cc, ymax + 3.*cc)
        ax.plot(strassex, strassey)
    
        line1, = ax.plot([], [], 'o-', lw=0.5)
        line2, = ax.plot([], [], linestyle = '--')
        line3, = ax.plot([], [], linestyle = '--')
        line4, = ax.plot([], [], 'bo', markersize=5) # the dot on the disc, to show it is rotating
        
        elli = patches.Circle((x1[0], y1[0]), radius = r1, fill=True, color='red', ec='black')
        ax.add_patch(elli)
    
        def animate(i):        
            ax.set_title('running time {:.2f} sec, friction ={}'.format(times[i], reibung1), fontsize=15)
            
            elli.set_center((x1[i], y1[i]))
            elli.set_height(2.*r1)
            elli.set_width(2.*r1)
            elli.set_angle(np.rad2deg(resultat[i, 0]))
                           
            line1.set_data(x1[i], y1[i])                  # center of the ellipse
            line2.set_data(geradex, horizontal_store[i])  # dashed line to mark the contact point
            line3.set_data(vertikal_store[i], geradey)    #            dto. 
            line4.set_data(x2[i], y2[i])
            return line1, line2, line3, line4,
    
        anim = animation.FuncAnimation(fig, animate, frames=len(times),
                                       interval=1000*max(times) / len(times),
                                       blit=True)
        plt.close(fig)
        return anim
    
    anim = animate_pendulum(times, Dmcx, Dmcy, Dmcox, Dmcoy)
    print('it took {:.3f} sec to run the program, before HTML(..)'.format(time.time() - start))
    HTML(anim.to_jshtml())    # needed, when run on an iPad, I know no other way to do it. It is SLOW!
 
```

**Comment by Peter230655** [comment]
Playing around some more, and with luck, I may have found something.
N			inertial frame
A2			fame fixed to homegenious disc
P0			point fixed in N
Dmc		center of the disc
CP			contact point of disc with the street
gesamt	models the street, a line in the X/Y plane

I did (all this is in the program I attached)
CP.set_pos(P0, x * N.x + gesamt * N.y)
CP.set_vel(N, auxx * N.x + auxy * N.y)
As CP is momentarily at rest in A2, I used v2pt_theory to get the speed of Dmc.
This works fine as long as the street is flat, that is the disc moves horizontally, but I could not get it to work on my uneven street.

So, now I tried: 
Dmc.set_vel(N, Dmc.pos_from(P0).diff(t, N) + auxx * N.x +auxy * N.y) [I added the virtual speeds auxx, auxy, to get the reaction forces.]
This worked fine, the total energy is constant up to numerical errors.

Since _v2pt_theory works fine when the road is flat_ (and in this case, subsequent CPs are also at different places), but it does _not seem to work when my street is uneven,_ I wonder whether there is an _issue with v2pt_theory_, or whether I applied it wrongly?

To summarize:
The only change I made to the program attached was to replace
Dmc.v2pt_theory(CP, N, A2)		with
Dmc.set_vel(N, Dmc.pos_from(P0).diff(t, N) + auxx * N.x +auxy * N.y)

and all seems fine, in the sense that the total energy is constant.

NB:
1.
I have a similar model in 3D. The same change fixed that program, too.

Any hints are greatly appreciated!
